playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml

```diff
@@ -567,46 +572,49 @@
   - role: openshift_examples
     openshift_examples_import_command: replace
   pre_tasks:
-  - name: Check for default router
+  - name: Collect all routers
     command: >
-      {{ oc_cmd }} get -n default dc/router
-    register: _default_router
+      {{ oc_cmd }} get pods --all-namespaces -l 'router' -o json
+    register: _all_routers
     failed_when: false
     changed_when: false
 
   - name: Check for allowHostNetwork and allowHostPorts
-    when: _default_router.rc == 0
+    when: _all_routers.rc == 0
     shell: >
       {{ oc_cmd }} get -o yaml scc/privileged | /usr/bin/grep -e allowHostPorts -e allowHostNetwork
     register: _scc
 
   - name: Grant allowHostNetwork and allowHostPorts
     when:
-    - _default_router.rc == 0
+    - _all_routers.rc == 0
     - "'false' in _scc.stdout"
     command: >
       {{ oc_cmd }} patch scc/privileged -p
       '{"allowHostPorts":true,"allowHostNetwork":true}' --api-version=v1
 
   - name: Update deployment config to 1.0.4/3.0.1 spec
-    when: _default_router.rc == 0
+    when: _all_routers.rc == 0
     command: >
-      {{ oc_cmd }} patch dc/router -p
+      {{ oc_cmd }} patch dc/{{ item['labels']['deploymentconfig'] }} -p
       '{"spec":{"strategy":{"rollingParams":{"updatePercent":-10},"spec":{"serviceAccount":"router","serviceAccountName":"router"}}}}'
       --api-version=v1
+    with_items: (_all_routers.stdout | from_json)['items'] | oo_pods_match_component(openshift_deployment_type, 'haproxy-router') | oo_select_keys_from_list(['metadata'])
```

Comment:
@detiber, @abutcher, I really struggled to make this do the right thing and right now there's a lot of duplicate logic.  What would be the best way to save the matching pods to another variable that could be reused?  I was probably doing something wrong at the time but the last time I tried using set_fact it was messing things up.  It that should work for a data structure like this then I can try again.